### PR TITLE
Add GitHub Action to trigger deploys on PR label assignment [720]

### PR DESCRIPTION
**Pull request by jtwillis92**
### Summary of Changes
Enables manually triggered deploys through CircleCI by label assignment on GitHub PRs.

* Adds a GitHub Action workflow that runs on the `labeled` event for Pull Requests in order to make a cURL call to the Circle CI API to trigger a `deploy` job. This workflow only executes if the tag name is explicitly `Deploy with CircleCI` and if all status checks are passing in CI for the PR.
* Updates the CircleCI config to create a `DOCKER_TAG` environment variable which is the branch name with the literal character `/` replaced with `-`. This variable is then used to define the name of the image tag within DockerHub that is used for deployments. This change was necessary because the `/` is an invalid character for docker tags and we were previously only using the branch name to define tags - since many of our branch names include the `/` character this would be rejected for those branches without the change.
* Updates README file to include how to manually trigger deployments and the process for setting that up on other repositories.

Addresses https://github.com/raft-tech/TANF-app/issues/720

### How to Test
* Assign the `Deploy with CircleCI` tag to this PR (if already applied, remove it first to trigger another deploy)
* Visit the [GitHub Actions](https://github.com/raft-tech/TANF-app/actions) tab and ensure that a new deploy job was started
* Click in to the latest execution of the workflow and visit the link shown in `build_url` as seen below:

circleci_job_url
* Check that the deployment completes and shows in the PR as seen below:

Screenshot from 2021-03-11 16-04-05

* If the PR status checks are failing or incomplete the deployment will be skipped as shown below:

circleci_action_skipped_state

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] Contributors with `triage` or higher access to the repo can trigger a deploy to `Development` on demand from a GitHub PR by applying the label `Deploy with CircleCI`
  - [x] README is updated, to explain how to trigger a deploy on demand
  - [x] Documented process for communicating deploys to the team
  - [ ] Stretch goal: Find a way to link the PR in the CircleCI deploy history so we can see what is on the dev site at the moment

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A Infrastructure changes

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A no frontend changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

**Comment by jtwillis92** [comment]
@carltonsmith this PR is blocked pending the current testing being done on the Development site. Once that is complete we can test deploying this branch against Dev.

**Comment by jtwillis92** [comment]
> @jtwillis92 I see this kicks off the deploy, but not the tests. Is there a way for it to run the whole CircleCI process?

It looks like it would be possible to add an `if` conditional expression to the Github Action that prevents the deploy job from being initiated if the required status checks aren't passing. However, that information isn't given to the action with the `pull_request.labeled` event. [Ref](https://docs.github.com/en/rest/reference/pulls#get-a-pull-request)

Instead, we'd have to get the `statuses_url` from the PR output we do have access to and add another job step that calls that URL to get the results of status checks. From there we would conditionally move on if those checks were successful and call the deploy.

@carltonsmith do you think this is something that should be addressed with the initial release or a new ticket?

**Comment by carltonsmith** [comment]
> > @jtwillis92 I see this kicks off the deploy, but not the tests. Is there a way for it to run the whole CircleCI process?
> 
> It looks like it would be possible to add an `if` conditional expression to the Github Action that prevents the deploy job from being initiated if the required status checks aren't passing. However, that information isn't given to the action with the `pull_request.labeled` event. [Ref](https://docs.github.com/en/rest/reference/pulls#get-a-pull-request)
> 
> Instead, we'd have to get the `statuses_url` from the PR output we do have access to and add another job step that calls that URL to get the results of status checks. From there we would conditionally move on if those checks were successful and call the deploy.
> 
> @carltonsmith do you think this is something that should be addressed with the initial release or a new ticket?

It depends on how much of a lift you think it is. Our commitment is for 1 or 2 point changes to be made in the initial PR, but if anything larger than that comes out of it, we can create a new issue and address down the road.

**Comment by jtwillis92** [comment]
> It depends on how much of a lift you think it is. Our commitment is for 1 or 2 point changes to be made in the initial PR, but if anything larger than that comes out of it, we can create a new issue and address down the road.
Ended up getting it added with this PR, updated now.

**Review by carltonsmith — APPROVED** [review]
LGTM!

**Comment by carltonsmith** [comment]
there is currently [a problem with CircleCi](https://status.circleci.com/) but once they get it fixed, we'll get the checks cleaned up

**Comment by carltonsmith** [comment]
> there is currently [a problem with CircleCi](https://status.circleci.com/) but once they get it fixed, we'll get the checks cleaned up

We re-ran everything this morning and everything looks better now. This is ready for your review @ADPennington

File: README.md [hunk 1]
```diff
@@ -82,3 +82,44 @@ Development | https://tdp-frontend.app.cloud.gov/ | `raft-review` in [Raft fork]
...
+This works using a [GitHub Action](https://docs.github.com/en/actions/quickstart) that runs every time a label is assigned to a PR.
+If the assigned label matches the string defined above a cURL request is made to CircleCI to initiate a deploy job for the given PR's branch.
+
+Which deployment environment within Cloud.gov the deploy job targets depends on the branch name as follows:
+* Branch name `staging` => deploys to Gov Staging **(once it exists)
+* Branch name `raft-tdp-main` => deploys to Vendor Staging
+* All other branches deploy to the Development environment
+
+Note that the Production environment is omitted above, since `main` is a protected branch commits can't be made directly
+to it so there is no path to be able to deploy straight to Production from labels on PRs.
```
Comment (ADPennington) [inline README.md:99]: 👍🏾 👍🏾

**Review by ADPennington — APPROVED** [review]
### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
Create a list of functional outcomes that must be achieved to complete this issue

 + [x] Contributors with triage or higher access to the repo can trigger a deploy to Development on demand from a GitHub PR by applying the label Deploy to Development
 + [x] README is updated, to explain how to trigger a deploy on demand
 + [x] Documented process for communicating deploys to the team


### Deliverable 2: Tested Code

_n/a_

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

_n/a_ 

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] implementation of github actions change well-documented

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_adequately tested the success and failure/pending routes. _

**Pull request merged**